assets/app/scripts/services/data.js

```diff
@@ -815,59 +807,74 @@ DataService.prototype.createStream = function(resource, name, context, opts, isR
     if (!events) {
       return false;
     }
 
     if (isTooManyRecentEvents(events)) {
-      Logger.log("Too many websocket open or close events for resource/context in a short period", resource, context, events);
+      Logger.log("Too many websocket open or close events for resource/context in a short period", key, events);
       return true;
     }
 
     if (isTooManyConsecutiveCloses(events)) {
-      Logger.log("Too many consecutive websocket close events for resource/context", resource, context, events);
+      Logger.log("Too many consecutive websocket close events for resource/context", key, events);
       return true;
     }
 
     return false;
   };
 
-  DataService.prototype._uniqueKeyForResourceContext = function(resource, context) {
-    // Note: when we start handling selecting multiple projects this
-    // will change to include all relevant scope
-    if (resource.equals("projects") || resource.equals("projectrequests")) { // when we are loading non-namespaced resources we don't need additional context
-      return resource;
-    }
-    else if (context.namespace) {
-      return resource.toString() + "/" + context.namespace;
-    }
-    else if (context.project && context.project.metadata) {
-      return resource.toString() + "/" + context.project.metadata.name;
-    }
-    else if (context.projectName) {
-      return resource.toString() + "/" + context.projectName;
-    }
-    else {
-      return resource.toString();
-    }
+
+  // will take an object, filter & sort it for consistent unique key generation
+  // uses encodeURIComponent internally because keys can have special characters, such as '='
+  var paramsForKey = function(params) {
+    var keys = _.keysIn(
+                  _.pick(
+                    params,
+                    ['fieldSelector', 'labelSelector'])
+                ).sort();
+    return _.reduce(
+            keys,
+            function(result, key, i) {
+              return (i < (keys.length-1)) ?
+                      result + key + '=' + encodeURIComponent(params[key]) + '&' :
+                      result + key + '=' + encodeURIComponent(params[key]);
+            }, '?');
+
+  };
+
+  // - proxies to _urlForResource to generate unique keys
+  // - ensure namespace if available
+  // - ensure only witelisted url params used for keys (fieldSelector, labelSelector) via paramsForKey
+  //   and that these are consistently ordered
+  // - NOTE: does not use the 'isWebsocket' bool as our watch/list operations should respond with the
+  //         same data from cache if the rest of the url (key) matches
+  DataService.prototype._uniqueKeyForResourceContext = function(resource, name, context, params) {
+    var ns = context && context.namespace ||
+             _.get(context, 'project.metadata.name') ||
+             context.projectName;
+    return this._urlForResource(resource, name, context, null, angular.extend({}, {}, {namespace: ns})).toString() + paramsForKey(params);
```

Comment:
maybe update the comment to say calling functions should not assume a valid URL is being returned, the only guarantee is that it is a unique key